File: vm/cpp/src/Bus.cpp
#include "../header/Bus.h"
#include "../header/CPU.h"
#include "../header/State.h"
#include <exception>

map<short, Device> Bus::devices;

void Bus::write(const short address, const int data) const
{
    auto dev = this->find(address);
    dev->write(address - dev->get_addr(), data);
}

int Bus::read(short address)
{
    throw new std::exception(&"access denied: "[address]);
}

void Bus::add(const Device* dev) const
{
    devices.insert(make_pair(dev->get_addr(), *dev));
}

Device* Bus::find(int address) const
{
    return &devices.at(static_cast<short>(address));
}

void Bus::setup(CPU* cpu, State* state)
{
    this->cpu = cpu;
    this->state = state;
}

File: vm/rust/src/Bus.rs
#[derive(Debug)]
pub struct Bus {
    pub state: State
}
pub struct Device {
    pub address: u32
}
impl Device {
    pub fn write(&self, address: u32, data: u32)
    {

    }
}
impl Bus {
    pub fn new(&st: State) -> Bus
    {
        Bus { state: st }
    }
    pub fn find(&self, address: u32) -> Device
    {
        Device { address: 0x1 }
    }
}